libobs-opengl/gl-egl-common.c

```diff
@@ -0,0 +1,262 @@
+/******************************************************************************
+    Copyright (C) 2020 by Georges Basile Stavracas Neto <georges.stavracas@gmail.com>
+
+    This program is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 2 of the License, or
+    (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+******************************************************************************/
+
+#include "gl-egl-common.h"
+
+#if defined(__linux__)
+
+#include <linux/types.h>
+#include <asm/ioctl.h>
+typedef unsigned int drm_handle_t;
+
+#else
+
+#include <stdint.h>
+#include <sys/ioccom.h>
+#include <sys/types.h>
+typedef int8_t __s8;
+typedef uint8_t __u8;
+typedef int16_t __s16;
+typedef uint16_t __u16;
+typedef int32_t __s32;
+typedef uint32_t __u32;
+typedef int64_t __s64;
+typedef uint64_t __u64;
+typedef size_t __kernel_size_t;
+typedef unsigned long drm_handle_t;
+
+#endif
+
+typedef void(APIENTRYP PFNGLEGLIMAGETARGETTEXTURE2DOESPROC)(
+	GLenum target, GLeglImageOES image);
+GLAPI PFNGLEGLIMAGETARGETTEXTURE2DOESPROC glad_glEGLImageTargetTexture2DOES;
+typedef void(APIENTRYP PFNGLEGLIMAGETARGETRENDERBUFFERSTORAGEOESPROC)(
+	GLenum target, GLeglImageOES image);
+GLAPI PFNGLEGLIMAGETARGETRENDERBUFFERSTORAGEOESPROC
+	glad_glEGLImageTargetRenderbufferStorageOES;
+static PFNGLEGLIMAGETARGETTEXTURE2DOESPROC glEGLImageTargetTexture2DOES;
+
+/* copied from drm_fourcc.h */
+
+#define fourcc_code(a, b, c, d)                                \
+	((__u32)(a) | ((__u32)(b) << 8) | ((__u32)(c) << 16) | \
+	 ((__u32)(d) << 24))
+#define DRM_FORMAT_XRGB8888        \
+	fourcc_code('X', 'R', '2', \
+		    '4') /* [31:0] x:R:G:B 8:8:8:8 little endian */
+
+static bool init_egl_image_target_texture_2d_ext(void)
+{
+	if (!glEGLImageTargetTexture2DOES) {
+		glEGLImageTargetTexture2DOES =
+			(PFNGLEGLIMAGETARGETTEXTURE2DOESPROC)eglGetProcAddress(
+				"glEGLImageTargetTexture2DOES");
```

Comment:
Im pretty sure mixing OpenGL ES with OpenGL is undefined behavior. And we dont request an OES context so this seems even more suspect. Is this to mean that there is no OpenGL support for dmabuf's?

Reply:
An nvidia employee suggested using this extension instead: https://www.khronos.org/registry/OpenGL/extensions/EXT/EXT_EGL_image_storage.txt

Reply:
This extension seems too recent, but it might be interesting to investigate it. The fact that it is made by NVidia doesn't make me comfortable though -- is this yet another NVidia-only extension in disguise?

@kkartaltepe looks like it is safe to use this function though, as per https://forums.developer.nvidia.com/t/can-i-use-an-eglimage-to-copy-between-gles-and-gl-in-the-same-process/67646/9

Reply:
"OES context" is not a thing, I think you mean a GL ES context? Since OES namespaced extensions are generally exclusive to GL ES.

If you check the extension strings advertised as suppoted for the whatever GL context you have (and you MUST always do this), and it says that `GL_OES_EGL_image` extension is supported, then you can just believe that. It is a driver bug if it is not actually true, not your problem. So if the extension is listed supported, feel free to make use of it.

Until I heard of `GL_EXT_EGL_image_storage` I thought there was no extension for using `EGLImage`s in desktop OpenGL (`GL_OES_EGL_image` is written against GL ES), which is just silly, so it's no wonder if some desktop GL implementations implemented it.

I think mixing desktop OpenGL and GL ES in the same process used to be bad, because the two libraries contained some same-named symbols and you'd get a call into the wrong one, exploding. I believe that is no longer true: Mesa has long (always?) had both libraries dispatch similarly anyway, so it didn't matter, and now we have glvnd which ensures calling the right implementation library anyway since you are allowed to mix implementations (like Mesa and NVIDIA drivers) in the same process for the same API even.

I've heard that Glamor (Xorg) is using `GL_OES_EGL_image` on desktop OpenGL as well.

Just make sure you've checked the extension string, and don't create new `EGLDisplay`s or GL contexts for things you want to use in existing displays or contexts.

Reply:
In any case, the point about not requesting the proper context is spot on, and I've added a patch fixing it. Thanks!

Reply:
I believe there is some confusion, when using eglcontexts you must explicitly bind the api you want to use. Either OpenGL or OpenGL ES. We bind OpenGL when making these EGLContexts current so drivers only have to support OpenGL calls and expecting them to support OpenGL ES seems suspect.
https://github.com/obsproject/obs-studio/pull/3338/files#diff-88897ab4e9cc569d06e52d9269d2634aR105

Now it was changed to request an OpenGL ES compatible context but is binding OpenGL which seems wrong as well (I believe its fine to ask for a context that supports both here). But what state is actually shared across the APIs? Im not sure, would just rebinding the context as OpenGL ES and using OpenGL textures work? (Probably since this works).

Reply:
EXT_EGL_image_storage seems like the only existing extension to import the EGLImage within OpenGL without resorting to OpenGL ES, but it would also require us to bump GL version to 4.2.

An alternative might be to just say we will use OpenGL ES instead on wayland but that may break any OpenGL specific code written elsewhere in OBS (Maybe thats no-where but the xcompcap plugin which doesnt work on wayland anyway).

Reply:
Excellent point. Unfortunately, switching to EGL_OPENGL_ES_API breaks everything. Drawing seems to fail with

```
error: glClear failed, glGetError returned GL_INVALID_OPERATION(0x502)
error: device_clear (GL) failed
error: device_draw (GL) failed
```

Needs more investigation

Reply:
I do not think it is possible to create one context for multiple rendering APIs, IOW do not try to create one context for OpenGL _and_ GL ES. You have to pick _one_ API.

EGL 1.5 specification, section 3.7.1 Creating Rendering Contexts: "contexts are specific to a single client API".

As I said, if the GL implementation advertises an extension, you can use it. It doesn't matter if a desktop OpenGL implementation advertises an OES extension. If it's there, it's usable, or you get to file a bug against the drivers. Please, do use the OES extension at least as a fallback when the EXT extension is not available.

For instance, Mesa on Intel Haswell does expose GL_OES_EGL_image and GL_OES_read_format on OpenGL 3.0.

Reply:
> It doesn't matter if a desktop OpenGL implementation advertises an OES extension.

Given OES extensions are written against OES specification the actual "correctness" of such an implementation is pretty suspect. (As far as one can even talk about correctness under OpenGL)

Though weirdly enough the GL_OES_EGL_image is written against the OpenGL (and not the OpenGL ES) spec even though it claims dependence on OpenGL ES, and yea, I see it being reported even under GLX. So I think you are right as long as the extension is reported it should work even though its pretty wild for khronos to just say yea lets add OES functions from OpenGL.

Reply:
Thanks everyone for your comments. I've reimplemented this check to check for the `GL_OES_EGL_image`, then checking if the function pointer exists. I believe this makes it conformant with what has been said so far, that as long as it's reported to be supported, then we can assume it's actually supported.

Reply:
Checking the function pointer you get from `eglGetProcAddress()` or the GLX equivalent is not necessary nor sufficient. The extensions string is the only one that matters.